### ISSUE TITLE
Improve connection speed by running fewer commands

### ISSUE THREAD

**Issue by chayim** [redis/go-redis#2850]
Today, connecting to a Redis instance, involves running multiple commands, and parsing the response repeatedly, this includes usage such as [setinfo](https://redis.io/commands/client-setinfo), which itself is run twice. The example below illustrates, that it's possible to pass all connection configuration information within a pipeline. In turn, the HELLO call returns the version of Redis, which could perhaps be used later for other things.

This is marked as _good-first-issue_ not because it's necessarily easy to implement, inasmuch as it's a good entrypoint, to learning how connections to redis occur.

```
127.0.0.1:6379> multi
OK
127.0.0.1:6379(TX)> auth bob smurf
QUEUED
127.0.0.1:6379(TX)> select 5
QUEUED
127.0.0.1:6379[5](TX)> hello 3
QUEUED
127.0.0.1:6379[5](TX)> client setinfo lib-name predis
QUEUED
127.0.0.1:6379[5](TX)>  client setinfo lib-ver 55.5.0
QUEUED
127.0.0.1:6379[5](TX)> client tracking on optin
QUEUED
127.0.0.1:6379[5](TX)> CLIENT caching yes
QUEUED
127.0.0.1:6379[5](TX)> exec
1) OK
2) OK
3) 1# "server" => "redis"
   2# "version" => "7.2.3"
   3# "proto" => (integer) 3
   4# "id" => (integer) 26
   5# "mode" => "standalone"
   6# "role" => "master"
   7# "modules" => (empty array)
4) OK
5) OK
6) OK
7) OK
127.0.0.1:6379[5]> 
```

**Comment by prabhjotsingh18** [comment]
Hey @chayim, 
total newbie here, I would love to take this up, but I would require assistance. 
If you can help me with stating the point and blueprint of how this can be implemented, that would be great.

**Comment by lzakharov** [comment]
Hello! If I understood correctly, the task is to optimize Redis connection initialization method [*baseClient.initConn](https://github.com/redis/go-redis/blob/master/redis.go#L280). I'm not sure that it's possible to move `hello` command inside the pipeline, but it can be easily done for two `ClientSetInfo` calls.

**Comment by prabhjotsingh18** [comment]
Sure @chayim,  @ofekshenawa if you can guide me where to begin and what concepts I need to familiarise myself with.

**Comment by pglushko** [comment]
@prabhjotsingh18  are you still working on this?

**Comment by prabhjotsingh18** [comment]
hey @pglushko , 
i have no idea how to proceed with the issue, was waiting on @ofekshenawa for direction, if you want you can take up the bug.

**Comment by subhakundu** [comment]
@pglushko are you working on it? If not, I would like to work on this. I would like to know how connections are working in Redis.

**Comment by subhakundu** [comment]
@chayim and @ofekshenawa, I am not sure if anyone is working on this issue or not. If not, I would like to work on it to learn. I would like your guidance on where to start.

**Comment by KyleKincer** [comment]
It seems like this has already been the case since #326 or am I missing something? https://github.com/redis/go-redis/pull/326/files#diff-190fc9ceda3bbdbe723d493b8b99e2b9a0100c7d635bfe9b04246280bef6f67e

**Comment by joe-ton** [comment]
Hi @chayim, @ofekshenawa — I see there's been some discussion and uncertainty about the current state of this issue. I'm new to contributing to go-redis and I'd love to help here.

Before diving in, I noticed @KyleKincer pointed out this might have been addressed already in [#326](https://github.com/redis/go-redis/pull/326/files#diff-190fc9ceda3bbdbe723d493b8b99e2b9a0100c7d635bfe9b04246280bef6f67e). Could someone confirm if the optimization is still needed or clarify what's specifically missing?

Happy to contribute and eager to learn more about Redis connection handling in the process. Appreciate any guidance to get started!

**Comment by ndyakov** [comment]
The connection handshake is mostly done in a pipeline and this should be addressed (for the most part) nowadays. Feel free to open new issue if you have an idea for further improvements of the handshake.
